transmogrifier/sources/dspace_dim.py

```diff
@@ -0,0 +1,276 @@
+
+from transmogrifier.models import (
+    AlternateTitle,
+    Contributor,
+    Date,
+    Funder,
+    Identifier,
+    Note,
+    RelatedItem,
+    Rights,
+    Subject,
+    TimdexRecord,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class DSpace_dim:
+    def __init__(
+        self,
+        source: str,
+        source_base_url: str,
+        source_name: str,
+        input_records: Iterator[Tag],
+    ) -> None:
+        self.source = source
+        self.source_base_url = source_base_url
+        self.source_name = source_name
+        self.input_records = input_records
+
+    def __iter__(self) -> Iterator[TimdexRecord]:
+        return self
+
+    def __next__(self) -> TimdexRecord:
+        xml = next(self.input_records)
+        record = self.create_from_dspace_dim(
+            self.source, self.source_base_url, self.source_name, xml
+        )
+        return record
+
+    @classmethod
+    def create_from_dspace_dim(
+        cls, source: str, source_base_url: str, source_name: str, xml: Tag
+    ) -> TimdexRecord:
+        """
+        Args:
+            source: A label for the source repository that is prepended to the
+            timdex_record_id.
+            source_base_url: The base URL for the source system from which direct links
+            to source metadata records can be constructed.
+            source_name: The full human-readable name of the source repository to be
+            used in the TIMDEX record.
+            xml: A BeautifulSoup Tag representing a single DSpace record in
+            dim XML.
+        """
+        # Required fields in TIMDEX
+        source_record_id = cls.create_source_record_id(xml)
+        all_fields = xml.metadata.find_all("field")
+        all_titles = [f for f in all_fields if f.attrs["element"] == "title"]
+        main_title = [t for t in all_titles if "qualifier" not in t.attrs]
+        if len(main_title) != 1:
+            raise ValueError(
+                "A record must have exactly one title. Titles found for record "
+                f"{source_record_id}: {main_title}"
+            )
+        kwargs = {
+            "source": source_name,
+            "source_link": source_base_url + source_record_id,
+            "timdex_record_id": f"{source}:{source_record_id.replace('/', '-')}",
+            "title": main_title[0].string,
+        }
+
+        # Optional fields in TIMDEX
+        # alternate_titles, uses full title list retrieved for main title field
+        for alternate_title in [
+            t
+            for t in all_titles
+            if "qualifier" in t.attrs and t.attrs["qualifier"] == "alternative"
+        ]:
+            a = AlternateTitle(
+                value=alternate_title.string,
+                kind="AlternativeTitle",
+            )
+            kwargs.setdefault("alternate_titles", []).append(a)
+
+        # contributors
+        citation_creators = []
+        for creator in [f for f in all_fields if f.attrs["element"] == "creator"]:
```

Comment:
We can just pull all contributors here and use the qualifier value as the kind value. Also, all of the records in the whoas records fixture seem to use "author" as the qualifier for authors so we probably want to use that to identify citation contributors.